shared/runtime/tinyusb_helpers.c

```diff
@@ -52,3 +52,32 @@ void tud_cdc_line_state_cb(uint8_t itf, bool dtr, bool rts) {
 }
 
 #endif
+
+
+#if MICROPY_HW_USB_MSC_EXCLUSIVE_ACCESS
+#include "tusb.h"
+#include "extmod/vfs.h"
+#include "extmod/vfs_fat.h"
+static mp_sched_node_t mp_remount_sched_node;
+
+STATIC void tud_msc_remount_task(mp_sched_node_t *node) {
+    mp_vfs_mount_t *vfs = NULL;
+    for (vfs = MP_STATE_VM(vfs_mount_table); vfs != NULL; vfs = vfs->next) {
+        if (vfs->len == 1) {
+            const char *path_str = "/";
+            mp_obj_t path = mp_obj_new_str(path_str, strlen(path_str));
+            nlr_buf_t nlr;
+            if (nlr_push(&nlr) == 0) {
+                mp_vfs_umount(vfs->obj);
+                mp_vfs_mount_and_chdir_protected(vfs->obj, path);
+                nlr_pop();
+            }
+            break;
+        }
+    }
+}
+
+void tud_msc_remount(void) {
+    mp_sched_schedule_node(&mp_remount_sched_node, tud_msc_remount_task);
```

Comment:
At which state will tud_msc_remount be called? Only when the device is ejected? Or is it possible that the PC may remount the drive? I cannot do that with my Linux box. If the board has always go through a power cycle, then it's fine.

Reply:
Assuming that a board with MSC support will start in locally write protected mode. What happens if the board is running stand-alone? Can it tell that there is no MSC mounted? Then the drive must be mounted write enabled.

Reply:
It gets called when the device is ejected. It may be possible to remount a safely-removed/ejected drive, but I think that disconnects power from the port anyway, I will have to test to confirm. However, I think that's a very unlikely case that shouldn't be handled, not sure if it should remount as read-only after giving write access to the device.

Reply:
I'm chewing on the timing issue. At the moment, the board starts with write enabled, to be able to create a file system, if it does not exists. Then it mounts the flash drive, which must be write protected if MSC is used. It could decide based on the status of the IOCTL_STATUS call. But how long does that status change take?

Reply:
OK. So the tud_msc_start_stop_cb() is only called on eject.  The ud_msc_inquiry_cb() is called on start, as well as the tud_msc_capacity_cb(). So we could use these as indicator, that a PC is attempting to connect.
The tud_msc_test_unit_ready_cb() is called regularly.

Reply:
TIming wise: If I set breakpoints as vfs_mount, tud_msc_inquiry_cb() and tud_msc_capacity_cb(), vfs_mount() is called first, and the others a few seconds later. Not really surprising. So we have to start with readonly mode for the local access, and maybe enable write access later.

Reply:
Actually everything works fine. I ported that over to MIMXRT, with the change to set the initial state of tud_msc_ejected to true, and set it to false in the calls of tud_msc_inquiry_cb()  and tud_msc_capacity_cb().
Then, when powering the board but not connecting it to a PC, it starts fine with write access.
When connected to a PC, the disk mounts and the local writes are blocked, until the disk is ejected.
That looks like a usable safe state.

Reply:
On a further analysis, the call to tud_msc_remount() is not needed at all. I was wondering if the remount happens at all, set  a breakpoint and a print statement in it. At no result. And it works even fine without it.

Reply:
I missed the `mkfs` use from boot script (and from REPL too). It needs to work the first time before the volume is mounted, but should fail if MSC is using the blockdev. However, I'm not sure how to fix that, so for now `mkfs` will always work regardless of MSC state.

Note made some fixes to ioctl's, refresh to see the changes.

Reply:
To expand on the above, `ff.c` uses `IOCTL_INIT` first from `f_mkfs` (this shouldn't return `PROTECT` otherwise boot.py will fail) and for all other functions (like open, write etc..) it calls `find_volume` which in turn uses `IOCTL_STATUS` to check the disk status (here it's safe to return `PROTECT` when MSC is using the blockdev). USB is usually initialized early, before running `boot.py`, so by the time `mkfs` is called the disk will be protected (or read only).

Reply:
MP calls disk_ioctl() and mimxrt_flash_ioctl() for the status, and blocks when the status of the latter is not Zero. direct calls to bdev.writeblocks() still work, when the device is mounted.
It may be that the changes to vfs_fat_diskio.c are also not needed.

Reply:
> USB is usually initialized early, before running boot.py

Indeed, the call for tusb_init() is done before running _boot.py, but the attempt to mount the disk is much later.

Reply:
The `STATUS` ioctl is now implemented with a separate command (see updated PR and updated comment above). It must be this way so `disk_ioctl` returns `PROTECT` if it's called from `find_volume` but not when called from `f_mkfs` (this is explained more above too).

Reply:
> Indeed, the call for tusb_init() is done before running _boot.py, but the attempt to mount the disk is much later.

Yes exactly, if this is fixed somehow I can make `mkfs` fail if the MSC is using the blockdev (because `boot.py` will be done before USB is enabled).

Reply:
Note that I changed my version of msc_disk to have ejected=true as start state. And I mount the disk in _boot.py as writable. So initially the disk is writable, like for mkfs or when no USB connects. That means, that the remount as writable is not needed.

Reply:
> Yes exactly, if this is fixed somehow I can make mkfs fail if the MSC is using the blockdev (because boot.py will be done before USB is enabled).

mkfs is needed to create a file system on a fresh new device. I do test that here easily with the Teensy boards, where the teensy loader erases the flash on each firmware load. Normally this is a nuisance, but in the actual case helpful.

Reply:
> Note that I changed my version of msc_disk to have ejected=true as start state. And I mount the disk in _boot.py as writable. So initially the disk is writable

But if the device enumerates before `_boot.py` it will set `tud_msc_ejected` back to false. Note that it may be working for you because `mkfs` only uses `IOCTL_INIT` (which doesn't check `tud_msc_ejected`).

Reply:
Why is that worse than having `tud_msc_ejected` being false from the beginning. That's a different topic, that exists anyhow in the current state. With `tud_msc_ejected` being true as initial state, devices not connected to a PC will work fine.

Reply:
That's right it needs to be `true` for the case where devices are running standalone. However, I don't think it will fix the mkfs issue.

Reply:
No. If the USB registers before the file system is created (what will is see without a file system ?), then there is no file system. But that will only happen after connecting a device without a FAT file system, like a Teensy board after loading the firmware. As a workaround the device could be powered once with e.g. a smartphone charger. Then it will create the file system.

Reply:
Or you use the PC to format the drive.

Reply:
> If the USB registers before the file system is created (what will is see without a file system ?), then there is no file system.

Wouldn't it set `tud_msc_ejected` to false ? Then `boot.py` runs and `mkfs` will fail to create the filesystem (if it doesn't exist). I think this will create a race condition and/or more issues than it's worth. I think the right way to fix it, is to prepare the filesystem before USB is enable (like `stm32`) which is the logical thing to do anyway.

Reply:
I can move tusb_init() behind the call to run _boot.py. That's not a problem. I tried that already for the mimxrt port for a different reason. Booting up without a file system being created is not a problem. That's what I tried just now. I can still create the file system on the PC.

Reply:
Okay, I set the initial value of `tud_msc_ejected` to true (and it gets cleared in `tud_msc_inquiry_cb`), also fixed `disk_ioctl` to check if the disk is write protected in `IOCTL_INIT`, so `mkfs` is handled too. But I didn't change `main.c` (the order of tusb_init/boot.py).

Reply:
> I can move tusb_init() behind the call to run _boot.py. That's not a problem

That would be best, as right now it probably depends on the timing to be right (i.e `boot.py` done first, the USB sets the `tud_msc_ejected` to false), however I will not change it in `rp2/main.c` because not sure if it will break something else, and if it's okay to call `tusb_init` every soft-boot or it should be called once etc..

Reply:
> if it's okay to call tusb_init every soft-boot or it should be called once etc..
 You can enforce a once only call with a state flag.

Reply:
I tried the "usb before boot.py" situation by adding a 5s delay into _boot.py before mounting. Nothing dramatic happens. There is just a drive without a file system. And on the next USB coennct, everything is fine.

Reply:
I missed your change in vfs_fat_diskio.c. WIth that, mkfs fails. But again without dramatic consequences. I liked the previous state more, because it was kind of self-healing.

Reply:
Okay how about now ? Now filesystem is created/set up before USB is enabled, and everything should be good.

Reply:
Yes, I tried a similar set-up, just with a local flag. Still not sure if that is a must have.
Another topic: what´s about dropping the remount after eject. I did not see any difference between doing it or not.

Reply:
If the host made any changes to the filesystem and/or fs index, they will not be seen by the device unless it's remounted, and the device could easily corrupt the filesystem.

Reply:
I do not know if the oofatfs does any kind of caching or buffering. If not, a remount does not add anything- At least the remount does not hurt. With both variants, with and w/o remount, i had rare cases were a change in the file system mad by the PC was visible until a soft reboot of the board. It were most of the time files with zero length (e.g. created with `touch` on the PC).

Reply:
I have no idea, maybe the writes from PC were not flushed ? Anyway, I'll clean up this PR tomorrow.

Reply:
I had not included you latest version of the remount code. But that does not help either. It looks as if the remount does not always happen. The error can be reproduced with some regularity using zero length files removed from the PC side.
If i create a file at the PC in sh using `touch test` followed by `sync`, it gets visible at the board even without ejecting the drive. If I then remove the file with `rm test`, followed by `sync`, the file stays visible. Even unmounting does not change the picture. After soft reset the file will not be displayed any more. So the file system WAS changed. Same behavior with Windows.

Reply:
Just to be sure, so you're saying I should do this:
* Create a file from PC with touch: `touch /mnt/board/test && sync`
* List dir `os.listdir()` from micropython.
* Remove the file from PC, `rm /mnt/board/test && sync`
* Eject the disk
* Softreboot `ctrl+d`
* List dir `os.listdir()` from micropython, the file is still there ?

Reply:
With the softreboot the file is gone. test without the soft reboot.
But I found a possible reason at my set-up: In tud_msc_remount_task() you have a check for `len == 1`. That is the length of the mount point name. On my set-up, the length is 6. So the remount never happens.
Changing the it to len == 6, the code is executed, and I can single-step it in the debugger. Then I see a strange phenomenon: while single stepping though unmount, remount, .., the PC still shows: "Unmount in progress, please do not remove the drive". Only after the remount function is finished, this message disappears. So it looks like in addition to the non-matching length check, the remount at the board was too early.

Reply:
> So it looks like in addition to the non-matching length check, the remount at the board was too early.

maybe a false alarm: The PC/USB simple code waits for  tud_msc_start_stop_cb() to finish.

Reply:
Yes maybe because you're single stepping the function. The remount is scheduled, so this function returns immediately. Anyway, the mount point of the MSC filesystem should be defined somewhere, or even better if it's detected dynamically somehow, but I'm not sure how to do that.

Reply:
I single stepped in the tud_msc_remount_task() function. So I know it is called.
At the moment, the internal flash is the first mounted file system. I do not know if we can reliably assume that this is always the case.

Reply:
If we do not find a good solution, we could at least defer opening the file system for write until the next soft boot. That's at least safe.

Reply:
If I try to do a umount and mount in REPL, I see a problem. The umount/mount seems to create new objects, and the old ones must be used anymore and show the wrong data.